rotkehlchen/chain/ethereum/modules/uniswap/v3/decoder.py

```diff
@@ -57,93 +69,245 @@ def _maybe_decode_v3_swap(  # pylint: disable=no-self-use
+            decoded_events: list[HistoryBaseEntry],
+    ) -> Optional[SwapData]:
+        """
+        Decode a swap of ETH to a token. Such swap consists of 3 events:
+        1. Sending ETH to the router.
+        2. If there is a rfund, receiving ETH from the router.
+        3. Receiving tokens from the router.
+        """
+        send_eth_event = decoded_events[1]
+        refund_eth_event = decoded_events[2] if decoded_events[2].asset == A_ETH else None
+        receive_token_event = decoded_events[2] if refund_eth_event is None else decoded_events[3]
+        from_amount = send_eth_event.balance.amount
+
+        if refund_eth_event is not None:
+            from_amount -= refund_eth_event.balance.amount
+
+        return SwapData(
+            from_asset=A_ETH,
+            from_amount=from_amount,
+            to_asset=receive_token_event.asset,
+            to_amount=receive_token_event.balance.amount,
+        )
+
+    def _decode_token_to_eth_swap(
+            self,
+            decoded_events: list[HistoryBaseEntry],
+    ) -> Optional[SwapData]:
+        from_asset: Optional[Asset] = None
+        from_amount = ZERO
+
+        # In this loop we find all the events that spend the token and sum them up.
+        # `0` event is the gas event, `1` event is the ETH receipt event, so we start with `2`.
+        for event in decoded_events[2:]:
+            if (
+                event.event_type == HistoryEventType.INFORMATIONAL and
+                event.event_subtype == HistoryEventSubType.APPROVE
+            ):
+                continue  # skip approval events
+
+            if (
+                event.event_type == HistoryEventType.TRADE and
+                event.event_subtype == HistoryEventSubType.SPEND and
+                event.counterparty in {CPT_UNISWAP_V2, CPT_UNISWAP_V3}
+            ) is False:
+                return None  # unexpected event
+
+            if from_asset is None:
+                from_asset = event.asset
+            elif from_asset != event.asset:
+                return None  # unexpected event
+
+            from_amount += event.balance.amount
+
+        return SwapData(
+            from_asset=from_asset,
+            from_amount=from_amount,
+            to_asset=A_ETH,
+            to_amount=decoded_events[1].balance.amount,  # the eth receipt event
+        )
+
+    def _decode_token_to_token_swap(
+            self,
+            decoded_events: list[HistoryBaseEntry],
+    ) -> Optional[SwapData]:
+        from_asset: Optional[Asset] = None
+        from_amount = ZERO
+        to_asset: Optional[Asset] = None
+        to_amount = ZERO
 
-        The swap method has as data the delta on the pool for the assets swapped so we make sure
-        to use signed integers to detect the amounts.
+        # In this loop we sum up events that spend `from_asset` and sum up
+        # events that receive the `to_asset`. `0` is the gas event, so we start with `1`.
+        for event in decoded_events[1:]:
```

Comment:
I wouldn't skip the first event. If we change something in the logic of how events get decoded we can end up with bugs here since you are making assumptions that are not really need. Why not just

```suggestion
        for event in decoded_events:
```

the gast event can be skipped in the first check you do.

Reply:
For better redability I tried to keep code here as explicit as possible. I think that if at any point we change the place of the gas event, it will require modifying many things. Or is redability not really an argument here?

Reply:
I don't see how it improves readability. Please change it

Reply:
same comment as before, for me it only gets confusing and in the other case you had to add a comment to explain why